src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -285,21 +517,95 @@ private static bool OperationUsesPreviewFeatures(OperationAnalysisContext contex
                 IBinaryOperation bOperation => bOperation.OperatorMethod,
                 IArrayCreationOperation arrayCreationOperation => (arrayCreationOperation.Type as IArrayTypeSymbol)?.ElementType,
                 ICatchClauseOperation catchClauseOperation => catchClauseOperation.ExceptionType,
+                ITypeOfOperation typeOfOperation => typeOfOperation.TypeOperand,
+                IEventAssignmentOperation eventAssignment => GetOperationSymbol(eventAssignment.EventReference),
                 _ => null,
             };
 
-        private static bool OperationUsesPreviewFeatures(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttribute)
+        private static bool TypeParametersHavePreviewAttribute(ImmutableArray<ITypeParameterSymbol> typeParameters, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttribute, [NotNullWhen(true)] out ISymbol? previewSymbol)
         {
-            return ProcessContainingTypePreviewAttributes(symbol, requiresPreviewFeaturesSymbols, previewFeatureAttribute) || ProcessPreviewAttribute(symbol, requiresPreviewFeaturesSymbols, previewFeatureAttribute);
+            foreach (ITypeParameterSymbol typeParameter in typeParameters)
+            {
+                ImmutableArray<ITypeSymbol> constraintTypes = typeParameter.ConstraintTypes;
+                foreach (ITypeSymbol constraintType in constraintTypes)
+                {
+                    if (SymbolIsAnnotatedAsPreview(constraintType, requiresPreviewFeaturesSymbols, previewFeatureAttribute))
+                    {
+                        previewSymbol = constraintType;
+                        return true;
+                    }
+                }
+            }
+
+            previewSymbol = null;
+            return false;
+        }
+
+        private static bool TypeArgumentsHavePreviewAttribute(ImmutableArray<ITypeSymbol> typeArguments, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttribute, [NotNullWhen(true)] out ISymbol? previewSymbol)
+        {
+            foreach (ITypeSymbol typeParameter in typeArguments)
+            {
+                if (SymbolIsAnnotatedAsPreview(typeParameter, requiresPreviewFeaturesSymbols, previewFeatureAttribute))
+                {
+                    previewSymbol = typeParameter;
+                    return true;
+                }
+            }
+
+            previewSymbol = null;
+            return false;
         }
 
-        private static bool ProcessPreviewAttribute(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol? previewFeatureAttribute)
+        private static bool SymbolIsAnnotatedAsPreview(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttribute)
```

Comment:
This is the most significant change. Previously I was only caching each symbol and a bool value. Now, I've changed this to cache each symbol and it's children in this call with a true value. I'm doing this as an optimization for deeply nested heirarachies where the Preview attribute is added close to the top.